### ISSUE TITLE
Some termcap variables are defined that should only be declared extern.

### ISSUE THREAD

**Issue by jvvv** [martinwguy/xvi#92]
The variables PC, BC, UP and ospeed are variables that are supposed to be provided by the terminal library. PC is the character used by tputs for padding. ospeed is the terminal ouput speed, which is used to determine the number of padding characters needed. BC and UP are used by tgoto as compensation strings for moving back or up. Setting these variables is the responsibility of the application, but each of these should be declared extern if they are to be declared at all (should be in termcap.h, though should have no effect if declared extern anyways).

**Comment by jvvv** [comment]
I have pushed an [issue92](https://github.com/jvvv/xvi/tree/issue92) branch with a candidate commit. My only misgiving is that I suspect it would be better to remove the declarations for these termcap variables all together and allow termcap.h to declare them for us.

**Comment by martinwguy** [comment]
That looks right, thanks. Yes, I changed it from BC to bc to be able to compile it with some linker that didn't merge global variables, but my fix was wrong. Thanks for spotting that!

Enabling "le" would be best as a separate commit, , but the following:
```
   if (BC == NULL)
	BC = LE;
```
doesn't look right. capability "bc" is null if the backspace character is the default of ^H, so we should use that not "left". There is boolean capability "bs" ("Control-H (8 dec.) performs a backspace"), so if that's false, "le" could be a useful substitute. I'm not sure from the manual what the termcap entry should be if the backspace character is something other than ^H... set both bs and bc?

Meanwhile, shall I drop the LE stuff and commit the scope fixes in your name?

**Comment by martinwguy** [comment]
My mistake again. I renamed xvi's BC to bc in commit 41536fc39 because
"<termcap.h> raises another problem: that it has extern char *BC;
   while tcap_scr.c has static char BC. Resolve by renaming xvi's one to bc."
so making it BC again would raise that type conflict again. I know the termcap library can make use of PC when you ask it to output other stuff (to create delays) but I can't think of any reason it would make use of BC, so our BC (or bc) being separate or the same makes no difference.  Is that right, or might we have to set termcap's BC for some reason?

However, fixing its bs/bc/le logic would be good.

termcap(5) says:
"The variables PC, UP and BC are set by tgetent to the terminfo  entry's
  data for pad_char, cursor_up and backspace_if_not_bs, respectively."
and
"UP is not used by ncurses.  PC is used in the tdelay_output function.
  BC is  used in the tgoto emulation."
(I guess tgoto emulation is what ncurses does using relative motions if there is no "cm" capability)

**Comment by jvvv** [comment]
The point about BC is that termcap defines it, so we must not redefine it.  My changes for BC, PC, UP, and ospeed were to make them external (or add as external, as I did with UP). But then, truly, they should not need to be declared at all, because they are declared in <termcap.h>. The termcap uses PC and ospeed for padding; and, BC and UP for movement compensations in  tgoto. It's our responsibility to see that these variables are properly set not declared or defined. So it is my mistake to even declare them as extern, unless of course there is some termcap.h that fails to declare them. I've been searching high and low for such, but even the earliest termcap/termlib I can find (circa 3BSD) has a working termcap.h.

**Comment by jvvv** [comment]
I've added a commit to my [issue92](https://github.com/jvvv/xvi/tree/issue92) branch that removes the termcap owned variables in tcap_scr.c unless AIX is defined, since for AIX we include curses.h. I added the termcap.h include to unix.c so that ospeed will get declared and added an extern ospeed decl for AIX.

**Comment by martinwguy** [comment]
Confirm issue92 branch compiles on netbsd.

On AIX 7.1 with makefile.aix I'm getting:
```
ld: 0711-317 ERROR: Undefined symbol: UP
ld: 0711-317 ERROR: Undefined symbol: BC
```

**Comment by jvvv** [comment]
One other thing. Your quote is from the ncurses manpage for termcap emulation in the terminfo portion of ncurses package. Traditional termcap expects these to be set by the
application. I have added another change that checks if BS is set, sets BC to "\b" if so, and otherwise sets BC to LE. I saw something very similar elsewhere in the wild, though I'm buggered if I can find it again.

**Comment by jvvv** [comment]
OK, for AIX, we may need the static variables then, or just ignore UP and BC for AIX all together.

**Comment by jvvv** [comment]
Ncurses and NetBSD's curses do a solid job with termcap emulation. Apparently, AIX doesn't see the need. I read something about the proprietary unices being less likely to follow portability, but I certainly can't back that up.

**Comment by martinwguy** [comment]
It would be interesting to consider using le if !bs, but only if there are any terminals that have le but not bs.
I've analysed /usr/share/misc/termcap (!) on netbsd, and it shows that
- there do exist terminals with no bs but with BC set (to ^H for two, to ^N for one). I think two of these terminal descriptions are wrong.
- there do exist 171 terminals with no "bs" or "bc" capability but with an "le" capability defined. All but 27 of these define it to ^H, so that gets us using ^H also with these terminals even though their termcap entries could be improved.

So definitely yes to using "le" if "bs" is false.

**Comment by martinwguy** [comment]
OK. Portability is a priority for xvi (which is why in parts it's in
#ifdef hell)

**Comment by jvvv** [comment]
It get's pretty hazy and it is almost like telling a story. In the original vi, that was how it was done, vi app code initialized PC, BC, UP, and ospeed. In the termcap library in 4.4BSD-Lite2, the UP and BC variables are statically declared in tgoto.c; the PC and ospeed variables statically declared in tputs.c. It is pretty much the same with the GNU termcap (version 1.3.1 from 2002, though 1.3 is from 1995 and not much different) and the linux rewrite by T. Tso (version 2.0.8, circa 1996).

As far as documentation, the 4.4BSD-Lite2 manual page termcap.3 says about the same as the NetBSD man page (not surprising since NetBSD had it's start based on 4.4BSD-Lite2, didn't it?). 

> 

     The tgoto() function returns a cursor addressing string decoded from cm
     to go to column destcol in line destline. It uses the external variables
     UP (from the up capability) and BC (if bc is given rather than bs) if
     necessary to avoid placing \n, ^D or ^@ in the returned string.  (Pro-
     grams which call tgoto() should be sure to turn off the XTABS bit(s),
     since tgoto() may now output a tab.  Note that programs using termcap
     should in general turn off XTABS anyway since some terminals use control-
     I for other functions, such as nondestructive space.)  If a % sequence is
     given which is not understood, then tgoto() returns (OOPS).

     The tputs() function decodes the leading padding information of the
     string cp; affcnt gives the number of lines affected by the operation, or
     1 if this is not applicable, outc is a routine which is called with each
     character in turn.  The external variable ospeed should contain the out-
     put speed of the terminal as encoded by stty(3).  The external variable
     PC should contain a pad character to be used (from the capability) if a
     null (^@) is inappropriate.

This isn't very explicit, but no where in the termcap libraries I've inspected are any of these variables set, only referenced. The GNU manual is quite specific, though:

> 

Initialization for Use of Termcap
=================================

   Before starting to output commands to a terminal using termcap, an
application program should do two things:

   * Initialize various global variables which termcap library output
     functions refer to.  These include `PC' and `ospeed' for padding
     (*note Output Padding::.) and `UP' and `BC' for cursor motion
     (*note tgoto::.).

Unfortunately, I haven't always kept notes on every link I have followed in trying to track down definitive instructions or at least firm suggestions on how apps that rely on termcap (or even terminfo) should initialize, as well as do's and don't's, etc. I've ordered Strang's 'Termcap and Terminfo' book as well as 'Programming with curses' nutshell book; should be here this week, as well as that debugging book.

**Comment by martinwguy** [comment]
> Before starting to output commands to a terminal using termcap,
> an application program should do two things:

Many thanks for researching that. At least whoever wrote the  GNU is clear about it!
So, does this sound right?:
- if we make use of the termcap library, we need to declare those four variables extern and initialise them ourselves from their capabilities (and ospeed from the "stty" terminal settings I guess) to make sure the library works OK. (Well, some implementations might work anyway, even without the inits, but for maximum portability, it is a compulsory thing to do)
- we don't *have* to use those variables for our own internal use (i.e. BC and UP), as they shouldn't be changed by termcap (well, except for the ones that initialise them themselves anyway), but we can if we like.
- AIX curses' termcap bits are so weak that they doesn't define these variables, so if we do use them ourselves inside xvi, we have to define them ourselves instead of just declaring them (char * instead of extern char *) .

My feeling says to separate the variables used by xvi from those used by termcap in this case, even if they seem similar, try to keep libtermcap happy by declaring/defining/initialising them best we can, but having our own variables for our own use, thereby helping decouple xvi from the differences between the various implementations, and giving it the best chance of compiling/running on other termcap implementations that we don't happen to have access to ourselves.

**Comment by martinwguy** [comment]
I've made a version of this on my branch issue92
https://github.com/martinwguy/xvi/pull/97
It compiles on Debian, NetBSD and AIX but I'd be grateful if you could check it against your understanding of the issues involved.

**Comment by jvvv** [comment]
I've looked over the changes. It makes sense that it compiles. I think, though, that there is likely a problem with only using the first character of these string capabilities. I realise that the current code already does this, but I'm not convinced it's correct. I think we are risking undefined behavior in the resulting output. I've been doing some checking on the length of the "bc", "do", "le", and "nd" capability strings in termcap and terminfo entries. While quite often these are just one character long, they are sometimes longer (i.e. escape sequences). By only using the first characters of escape sequences, I think we are inviting output bugs. I think I understand the intent in the original xvi code of only using the first characters: to speed up terminal output when making cursor movements that are often achieved with only one character (i.e. "bc" is often '^H', but I've also seen it as a sequence in the ncurses terminfo/termcap sources).

My opinion is that we should make attempts at increasing the terminal output performance in other ways: rework the update code to better utilise available capabilities, find places where we can avoid making cursor movements, simplify cursor movements when current position is known and we are moving in well defined patterns... To my way of looking at the topic, I think we would be safer in the long run to replace all instances that output a capability with only it's first character. The "pc" capability to PC variable relationship is the only one where this is guaranteed to be valid to only use the first character, but this is well defined in the documentation and PC is specifically used by tputs (and possibly tgoto) for padding.

One of the things I have considered is to expand the virtscr abstraction such that system specific screen/window update code can be overlayed in place of the existing portable code. That way, existing systems that we haven't optimised can default to the portable code. I haven't found time to start this, though I'm itching to get to it.

**Comment by martinwguy** [comment]
> only using the first characters of escape sequences
> I think we are risking undefined behavior in the resulting output.

It doesn't use the first character of escae sequences. Instead, it
only uses those capabilities at all if they are exactly one character
long, so a 2-char bc won't be used by xvi. The commented rationale is
that it makes the number-of-output-characters calculations easier, but
results in it not using a shorter sequence to move left by one place,
if it is available (and after all moving one char left is a fairly
common occurrence!)

The result on the screen seems correct, it's just just pessimal as far
as number of output characters goes.

> rework the update code to better utilise available capabilities

Absolutely.

> replace all instances that output a capability with only it's first character.

Or, rather, to use those capabilities including when they are more
than 1 char long.

The general problem, of finding the least number of output characters
to change one screen image into another, is actually very difficult as
there are so many possible solutions.
This is the problem "curses" faces, but xvi seems to compare what-is
and what-should-be is when updating the status line, if I remember
correctly.

When xvi was born, most people worked over 9600 baud serial lines -
almost 10 characters a second! - so less characters was critical but
these days most users are on directly-connected fast screens and xvi's
portability targets are Amiga, Atari ST, MSDOS and so on, machines
with attached screens but small memory. Of course, it should still be
usable over a slow connection but small code size and memory use are
more important.

However looking at xvi's character output, it moves here and there and
turns colours on and off without actually outputting anything before
moving on. It's not a defect, as such, but is worrying, as is issue
#72. It seems overly fond of repainting entire lines where a motion
would suffice.

> The "pc" capability to PC variable relationship is the only one where this is guaranteed to be valid
> to only use the first character

Yes, after all, "bc", despite it name, is a string capability.

> One of the things I have considered is to expand the virtscr abstraction
> such that system specific screen/window update code can be overlayed in
> place of the existing portable code. That way, existing systems that we
> haven't optimised can default to the portable code. I haven't found time to
> start this, though I'm itching to get to it.

Great!

**Comment by martinwguy** [comment]
Fixed by 63bcea8d
